Fix issues in launch_pager().

markus-oberhumer opened the pull request.

markus-oberhumer commented:
Hi Tavian,

first many thanks for your nice bfs tool!

Since the last 20 years or so I'm using `export LESS="-MM -i"` in work environment - and this does conflict with the default handling of `bfs --help`. Furthermore `LESS= bfs --help` also did not work.

So I've tried to fix that in this pull request - please have a look.

Cheers,
Markus

tavianator commented on parse.c:
```diff
@@ -2535,46 +2535,26 @@ static CFILE *launch_pager(pid_t *pid, CFILE *cout) {
 	bfs_spawn_destroy(&ctx);
 	return ret;
 
 fail_envp:
```
This label is empty now, might as well remove it

tavianator commented on spawn.c:
```diff
@@ -125,6 +128,16 @@ int bfs_spawn_addfchdir(struct bfs_spawn *ctx, int fd) {
 	}
 }
 
+int bfs_spawn_addputenv(struct bfs_spawn *ctx, char *s) {
```
I would make this take `const char *` and `strdup()` it

tavianator commented:
Also, do you mind adding this patch next time you push?  It should keep the FreeBSD build from hanging forever, since it doesn't have access to the necessary secrets when running a fork.  I thought GitHub did this automatically but it just runs it with empty strings.

```diff
diff --git a/.github/workflows/freebsd.yml b/.github/workflows/freebsd.yml
index 945be19..0939303 100644
--- a/.github/workflows/freebsd.yml
+++ b/.github/workflows/freebsd.yml
@@ -4,6 +4,8 @@ on: [push, pull_request]
 
 jobs:
   build:
+    if: github.repository_owner == "tavianator"
+
     runs-on: ubuntu-latest
 
     steps:
```

markus-oberhumer commented:
I've addressed your review request and pushed a new verions of my branch.  BTW, I could also squash the commits to make the history look nicer if you prefer.

As for git, I never had any issues with its pager handling - but I'll have a look at the git sources.

markus-oberhumer commented:
BTW, it seems that YAML syntax is not correct - please see https://github.com/tavianator/bfs/actions/runs/745604714 .

> Also, do you mind adding this patch next time you push? It should keep the FreeBSD build from hanging forever, since it doesn't have access to the necessary secrets when running a fork. I thought GitHub did this automatically but it just runs it with empty strings.
> 
> ```diff
> diff --git a/.github/workflows/freebsd.yml b/.github/workflows/freebsd.yml
> index 945be19..0939303 100644
> --- a/.github/workflows/freebsd.yml
> +++ b/.github/workflows/freebsd.yml
> @@ -4,6 +4,8 @@ on: [push, pull_request]
>  
>  jobs:
>    build:
> +    if: github.repository_owner == "tavianator"
> +
>      runs-on: ubuntu-latest
>  
>      steps:
> ```

tavianator commented:
> I've addressed your review request and pushed a new version of my branch. BTW, I could also squash the commits to make the history look nicer if you prefer.

Thanks!  I'd like the fixes folded into the relevant commits, but don't squash the whole thing.

> As for git, I never had any issues with its pager handling - but I'll have a look at the git sources.

Strange.  `LESS= git log` prints `ESC[33mcommit ...` for me.

> BTW, it seems that YAML syntax is not correct - please see https://github.com/tavianator/bfs/actions/runs/745604714 .

Ah oops, I think it has to be single quotes.  Didn't realize that would make a difference.

tavianator commented:
Also if you can s/supprt/support/g in 9fe44c83 that'd be good

markus-oberhumer commented:
Rebased and cleaned up. Please review but don't merge yet - still looking at git.

markus-oberhumer commented:
BTW, that FreeBSD/tailscale CI test is very slow (running for more than 15 minutes right now).

tavianator commented:
Thanks!  Looks good.

> BTW, that FreeBSD/tailscale CI test is very slow (running for more than 15 minutes right now).

It's slow because it doesn't have the secrets, so it stalls waiting for me to manually approve the tailscale connection.  I thought the `if:` would stop it from running but I think the condition is always true.  Can you try this instead?

```
if: github.event.pull_request.head.repo.full_name == github.repository
```

markus-oberhumer commented:
Sigh - you are right about `git`. I've just realized I have a `pager = less -FRX` setting in my ~/.gitconfig ...

And as closely following git is probably a *very* good idea (even if the `LESS= git log` behaviour is questionable) this pull request should probably get closed.

markus-oberhumer commented:
> if: github.event.pull_request.head.repo.full_name == github.repository

I committed this in my latest (squashed) version of this issue, and yes that seems to work!

markus-oberhumer commented:
I'm closing this fow now - will resubmit an improved version soon.
